File: leet-code/316. Remove Duplicate Letters/RemoveDuplicateLetters316_4.java
package com.example;

import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

public class RemoveDuplicateLetters316_4 {
    public String removeDuplicateLetters(String s) {
        Map<Character, Integer> counter = new HashMap<>();
        Map<Character, Boolean> existsInStack = new HashMap<>();
        Deque<Character> stack = new LinkedList<>();
        for (char c : s.toCharArray()) {
            counter.putIfAbsent(c, 0);
            counter.put(c, counter.get(c) + 1);
            existsInStack.putIfAbsent(c, false);
        }
        for (char c : s.toCharArray()) {
            counter.put(c, counter.get(c) - 1);
            if (existsInStack.get(c)) {
                continue;
            }
            while (!stack.isEmpty() && stack.peek() > c && counter.get(stack.peek()) > 0) {
                existsInStack.put(stack.pop(), false);
            }
            stack.push(c);
            existsInStack.put(c, true);
        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        return sb.reverse().toString();
    }
}

File: leet-code/316. Remove Duplicate Letters/RemoveDuplicateLetters316_4.kt
package com.example.algorithm

import java.util.LinkedList

class RemoveDuplicateLetters316_4 {
    fun removeDuplicateLetters(s: String): String {
        val counter = mutableMapOf<Char, Int>()
        val existsInStack = mutableMapOf<Char, Boolean>()
        val stack = LinkedList<Char>()
        for (c in s) {
            counter[c] = counter.getOrDefault(c, 0) + 1
            existsInStack[c] = false
        }
        for (c in s) {
            counter[c] = counter.getOrDefault(c, 0) - 1
            if (existsInStack[c] == true) {
                continue
            }
            while (stack.isNotEmpty() && stack.peek() > c && counter[stack.peek()]!! > 0) {
                existsInStack[stack.pop()] = false
            }
            existsInStack[c] = true
            stack.push(c)
        }
        val sb = StringBuilder()
        while (stack.isNotEmpty()) {
            sb.append(stack.pop())
        }
        return sb.reverse().toString()
    }
}
